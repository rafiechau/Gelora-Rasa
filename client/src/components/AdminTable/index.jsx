import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import classes from './style.module.scss';

const AdminTable = ({
  columns,
  data,
  onEdit,
  onDelete,
  showEditButton = true,
  showDeleteButton = true,
  editButtonMessageId = 'app_btn_view_details',
}) => (
  <div className={classes.responsiveTableContainer} data-testid="admin-table">
    <table className={classes.responsiveTable}>
      <thead>
        <tr>
          {columns.map((column) => (
            <th key={column.id}>
              <FormattedMessage id={column.messageId} defaultMessage={column.label} />
            </th>
          ))}
          <th>
            <FormattedMessage id="app_header_actions" />
          </th>
        </tr>
      </thead>
      <tbody>
        {data.map((item, index) => (
          <tr key={item.id}>
            {columns.map((column) => (
              <td key={`${item.id}-${column.id}`}>{item[column.id]}</td>
            ))}
            <td className={classes.actions} data-testid={`row-actions-${index}`}>
              {showEditButton && (
                <button
                  type="button"
                  onClick={() => onEdit(item)}
                  className={`${classes.btn} ${classes.btnDetail}`}
                  data-testid={`edit-button-${item.id}`}
                >
                  <FormattedMessage id={editButtonMessageId} />
                </button>
              )}
              {showDeleteButton && ( // Kondisional untuk menampilkan tombol delete
                <button
                  type="button"
                  onClick={() => onDelete(item.id)}
                  className={`${classes.btn} ${classes.btnDelete}`}
                  data-testid={`delete-button-${item.id}`}
                >
                  <FormattedMessage id="app_btn_delete" />
                </button>
              )}
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  </div>
);

AdminTable.propTypes = {
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  onEdit: PropTypes.func,
  onDelete: PropTypes.func,
  showEditButton: PropTypes.bool,
  showDeleteButton: PropTypes.bool,
  editButtonMessageId: PropTypes.string,
};

export default AdminTable;
