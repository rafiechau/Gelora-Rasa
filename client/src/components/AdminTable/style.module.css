.responsiveTableContainer {
  overflow-x: auto;
}
.responsiveTableContainer .responsiveTable {
  width: 100%;
  border-collapse: collapse;
  background-color: #f3f3f3;
  color: #333;
  font-size: 1rem;
}
.responsiveTableContainer .responsiveTable thead {
  background-color: #333;
  color: #fff;
}
.responsiveTableContainer .responsiveTable thead th {
  padding: 10px;
  text-align: left;
  border-bottom: 1px solid #ddd;
}
.responsiveTableContainer .responsiveTable tbody td {
  padding: 10px;
  text-align: left;
  border-bottom: 2px solid #ddd;
}
.responsiveTableContainer .responsiveTable tbody .actions {
  display: flex;
  flex-wrap: wrap;
  gap: 2px;
}
.responsiveTableContainer .responsiveTable tbody .actions .btn {
  padding: 10px 15px;
  margin-right: 5px;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  flex: 1;
}
.responsiveTableContainer .responsiveTable tbody .actions .btn-detail {
  background-color: #3498db;
  color: #fff;
}
.responsiveTableContainer .responsiveTable tbody .actions .btn-delete {
  background-color: #e74c3c;
  color: #fff;
}

@media screen and (max-width: 722px) {
  .responsiveTable tbody td:before {
    content: attr(data-label);
    font-weight: bold;
    margin-bottom: 5px;
  }
  .responsiveTable tbody td {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  }
  .responsiveTable tbody .actions {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  }
  .responsiveTable tbody .actions .btn {
    width: 100%;
    margin: 5px 0;
  }
}
@media screen and (max-width: 600px) {
  .responsiveTable thead {
    display: none;
  }
  .responsiveTable tbody td:before {
    content: attr(data-label);
    font-weight: bold;
    margin-right: 10px;
  }
  .responsiveTable tbody td {
    display: flex;
    width: 100%;
    box-sizing: border-box;
    padding: 10px 5px;
  }
  .responsiveTable tbody .btn {
    padding: 10px;
    font-size: 0.8rem;
  }
}
@media screen and (max-width: 840px) {
  .responsiveTable thead th {
    padding: 8px;
  }
  .responsiveTable tbody td {
    word-break: break-word;
  }
  .responsiveTable tbody .btn {
    width: 100%;
    padding: 8px 100px;
    font-size: 0.75rem;
    margin: 5px 0;
    width: auto;
    max-width: 100%;
  }
}
@media screen and (max-width: 900px) {
  .responsiveTable thead th {
    font-size: 0.7rem;
    padding: 8px;
  }
  .responsiveTable tbody td {
    font-size: 0.7rem;
    word-break: break-word;
  }
  .responsiveTable tbody .actions .btn {
    padding: 1px 3px;
    font-size: 0.5rem;
    margin: 5px 0;
    width: auto;
    max-width: 100%;
  }
}
@media screen and (max-width: 840px) {
  .responsiveTable thead th {
    padding: 8px;
  }
  .responsiveTable tbody td {
    word-break: break-word;
  }
  .responsiveTable tbody .actions .btn {
    padding: 8px;
    font-size: 0.75rem;
    margin: 5px 0;
    width: auto;
    max-width: 100%;
  }
}
@media screen and (max-width: 725px) {
  thead {
    display: none;
  }
  .responsiveTable tbody .actions {
    justify-content: space-between;
    gap: 8px;
  }
  .responsiveTable tbody .actions .btn {
    padding: 5px;
    font-size: 0.65rem;
  }
}
@media screen and (max-width: 722px) {
  .responsiveTable tbody .actions .btn {
    width: 100%;
    margin: 5px 0;
  }
}/*# sourceMappingURL=style.module.css.map */