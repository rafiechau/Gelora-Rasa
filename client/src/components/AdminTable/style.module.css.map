{"version":3,"sources":["style.module.scss","style.module.css"],"names":[],"mappings":"AAAA;EACI,gBAAA;ACCJ;ADAI;EACI,WAAA;EACA,yBAAA;EACA,yBAAA;EACA,WAAA;EACA,eAAA;ACER;ADDQ;EACI,sBAAA;EACA,WAAA;ACGZ;ADFY;EACI,aAAA;EACA,gBAAA;EACA,6BAAA;ACIhB;ADCY;EACI,aAAA;EACA,gBAAA;EACA,6BAAA;ACChB;ADCY;EACI,aAAA;EACA,eAAA;EACA,QAAA;ACChB;ADAgB;EACI,kBAAA;EACA,iBAAA;EACA,YAAA;EACA,kBAAA;EACA,eAAA;EACA,OAAA;ACEpB;ADAgB;EACI,yBAAA;EACA,WAAA;ACEpB;ADCgB;EACI,yBAAA;EACA,WAAA;ACCpB;;ADOA;EAGY;IACI,yBAAA;IACA,iBAAA;IACA,kBAAA;ECNd;EDQU;IACI,aAAA;IACA,sBAAA;IACA,uBAAA;ECNd;EDQU;IACI,aAAA;IACA,sBAAA;IACA,uBAAA;ECNd;EDOc;IACI,WAAA;IACA,aAAA;ECLlB;AACF;ADWA;EAEQ;IACI,aAAA;ECVV;EDaU;IACI,yBAAA;IACA,iBAAA;IACA,kBAAA;ECXd;EDaU;IACI,aAAA;IACA,WAAA;IACA,sBAAA;IACA,iBAAA;ECXd;EDaU;IACI,aAAA;IACA,iBAAA;ECXd;AACF;ADkBE;EAGU;IACI,YAAA;EClBd;EDuBU;IACI,sBAAA;ECrBd;EDuBU;IACI,WAAA;IACA,kBAAA;IACA,kBAAA;IACA,aAAA;IACA,WAAA;IACA,eAAA;ECrBd;AACF;AD0BA;EAGY;IACI,iBAAA;IACA,YAAA;EC1Bd;ED+BU;IACI,iBAAA;IACA,sBAAA;EC7Bd;EDgCc;IACI,gBAAA;IACA,iBAAA;IACA,aAAA;IACA,WAAA;IACA,eAAA;EC9BlB;AACF;ADqCA;EACI;IACI,YAAA;ECnCN;EDqCE;IACI,sBAAA;ECnCN;EDqCE;IACI,YAAA;IACA,kBAAA;IACA,aAAA;IACA,WAAA;IACA,eAAA;ECnCN;AACF;ADsCA;EACI;IACI,aAAA;ECpCN;EDsCE;IACI,8BAAA;IACA,QAAA;ECpCN;EDsCE;IACI,YAAA;IACA,kBAAA;ECpCN;AACF;ADuCA;EACI;IACI,WAAA;IACA,aAAA;ECrCN;AACF","file":"style.module.css"}