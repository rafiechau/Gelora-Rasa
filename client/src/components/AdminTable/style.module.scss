.responsiveTableContainer {
    overflow-x: auto;
    .responsiveTable{
        width: 100%;
        border-collapse: collapse;
        table-layout: fixed;
        background-color: $responsiveTableBackground;
        color: $color-white;
        font-size: 1rem;
        box-sizing: border-box;
        thead{
            background-color: darken($responsiveTableBackground, 10%);
            color: $color-white;
            th{
                padding: 10px;
                text-align: left;
                border-bottom: 1px solid #ddd;
               
            }
        }
        tbody{
            td{
                font-size: .8rem;
                padding: 10px;
                text-align: left;
                border-bottom: 2px solid #ddd;
            }
            .actions{
                // display: flex;
                // flex-direction: column;
                align-items: stretch;
                gap: 2px;
                .btn {
                    padding: 10px 15px;
                    margin-right: 5px;
                    border: none;
                    border-radius: 4px;
                    cursor: pointer;
                    flex: 1; 
                    min-width: 60px;
                }
                .btn-detail {
                    background-color: $button-primary;
                    color: $color-white;
                }
                  
                .btn-delete {
                    border: none;
                    background-color: $button-delete;
                    color: $color-white;
                }
            }
            
        }
    }
}

@media screen and (max-width: 722px) {
    .responsiveTable{
        tbody{
            td:before {
                content: attr(data-label);
                font-weight: bold;
                margin-bottom: 5px;
            }
            td {
                display: flex;
                flex-direction: column;
                align-items: flex-start; 
            }
            .actions {
                display: flex;
                flex-direction: column;
                align-items: flex-start;
                .btn {
                    width: 100%;
                    margin: 5px 0;
                  }
            }
        } 
    }
}

@media screen and (max-width: 600px) {
    .responsiveTable{
        thead {
            display: none;
        }
        tbody{
            td:before {
                content: attr(data-label);
                font-weight: bold;
                margin-right: 10px;
            }
            td{
                display: flex;
                width: 100%;
                box-sizing: border-box;
                padding: 10px 5px;
            }
            .btn {
                padding: 10px;
                font-size: 0.8rem;
              }
        }
    } 
  
    
  }

  @media screen and (max-width: 840px) {
    .responsiveTable{
        thead{
            th{
                padding: 8px;
            }
        }
        
        tbody{
            td{
                word-break: break-word;
            }
            .btn {
                width: 100%;
                padding: 8px 100px; 
                font-size: 0.75rem; 
                margin: 5px 0;
                width: auto;
                max-width: 100%;
            } 
        }
    }
}


@media screen and (max-width: 840px) {
    .responsiveTable thead th {
        padding: 8px;
    }
    .responsiveTable tbody td {
        word-break: break-word;
    }
    .responsiveTable tbody {
        margin-top: 5px;
        width: 100%;
    } 
    .btn{
        margin-top: 5px;
        width: 100%;
    }
}

@media screen and (max-width: 725px) {
    thead{
        display: none;
    }
    .responsiveTable tbody .actions {
        justify-content: space-between; 
        gap: 8px; 
    }
    .responsiveTable tbody .actions .btn {
        padding: 5px; 
        font-size: 0.65rem; 
    }
}


@media screen and (max-width: 1065px) {
    .btn{
        margin-top: 5px;
        width: 100%;
    }
}