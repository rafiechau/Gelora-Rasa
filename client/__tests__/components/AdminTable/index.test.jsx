import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import AdminTable from '@components/AdminTable';

jest.mock('react-intl', () => ({
  // eslint-disable-next-line react/prop-types
  FormattedMessage: ({ id }) => <span>{id}</span>,
}));

const columns = [
  { id: 'name', messageId: 'name', label: 'Name' },
  { id: 'age', messageId: 'age', label: 'Age' },
];

const data = [
  { id: '1', name: 'John Doe', age: '30' },
  { id: '2', name: 'Jane Doe', age: '25' },
];

const mockOnEdit = jest.fn();
const mockOnDelete = jest.fn();

describe('AdminTable Component', () => {
  it('should render the component correctly', () => {
    const { getByTestId, getByText } = render(
      <AdminTable columns={columns} data={data} onEdit={mockOnEdit} onDelete={mockOnDelete} />
    );

    // Check for the presence of the table
    expect(getByTestId('admin-table')).toBeInTheDocument();

    // Check if column headers are rendered
    columns.forEach((column) => {
      expect(getByText(column.id)).toBeInTheDocument();
    });

    // Check if data is rendered
    data.forEach((row) => {
      columns.forEach((column) => {
        expect(getByText(row[column.id])).toBeInTheDocument();
      });
    });
  });

  it('should handle edit button click', () => {
    const { getByTestId } = render(
      <AdminTable columns={columns} data={data} onEdit={mockOnEdit} onDelete={mockOnDelete} />
    );

    // Simulate click on the first edit button
    fireEvent.click(getByTestId(`edit-button-${data[0].id}`));
    expect(mockOnEdit).toHaveBeenCalledWith(data[0]);
  });

  it('should handle delete button click', () => {
    const { getByTestId } = render(
      <AdminTable columns={columns} data={data} onEdit={mockOnEdit} onDelete={mockOnDelete} />
    );

    // Simulate click on the first delete button
    fireEvent.click(getByTestId(`delete-button-${data[0].id}`));
    expect(mockOnDelete).toHaveBeenCalledWith(data[0].id);
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(
      <AdminTable columns={columns} data={data} onEdit={mockOnEdit} onDelete={mockOnDelete} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
