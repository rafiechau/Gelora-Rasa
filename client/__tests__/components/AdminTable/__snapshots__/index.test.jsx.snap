// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminTable Component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="responsiveTableContainer"
    data-testid="admin-table"
  >
    <table
      class="responsiveTable"
    >
      <thead>
        <tr>
          <th>
            <span>
              name
            </span>
          </th>
          <th>
            <span>
              age
            </span>
          </th>
          <th>
            <span>
              app_header_actions
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            John Doe
          </td>
          <td>
            30
          </td>
          <td
            class="actions"
            data-testid="row-actions-0"
          >
            <button
              class="btn btnDetail"
              data-testid="edit-button-1"
              type="button"
            >
              <span>
                app_btn_view_details
              </span>
            </button>
            <button
              class="btn btnDelete"
              data-testid="delete-button-1"
              type="button"
            >
              <span>
                app_btn_delete
              </span>
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Jane Doe
          </td>
          <td>
            25
          </td>
          <td
            class="actions"
            data-testid="row-actions-1"
          >
            <button
              class="btn btnDetail"
              data-testid="edit-button-2"
              type="button"
            >
              <span>
                app_btn_view_details
              </span>
            </button>
            <button
              class="btn btnDelete"
              data-testid="delete-button-2"
              type="button"
            >
              <span>
                app_btn_delete
              </span>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
